File: backend/src/NutriSoftware.Domain/Nutrientes.cs
namespace NutriSoftware.Domain;

/// <summary>Un nutriente de la tabla de composición.</summary>
/// <param name="Clave">Clave estable usada en los JSON de composición.</param>
/// <param name="Etiqueta">Nombre para mostrar.</param>
/// <param name="Unidad">Unidad por 100 g de alimento.</param>
/// <param name="Decimales">Decimales con los que se informa el total.</param>
public record Nutriente(string Clave, string Etiqueta, string Unidad, int Decimales);

/// <summary>
/// Catálogo de los 22 nutrientes de la tabla peruana de composición de
/// alimentos, en el orden en que se muestran.
///
/// Es la fuente única: el importador, el cálculo de totales y la exportación
/// recorren esta lista, así que agregar un nutriente es agregar una fila acá.
/// El frontend replica estas claves en <c>frontend/src/lib/nutrientes.ts</c>.
/// </summary>
public static class Nutrientes
{
    public static readonly IReadOnlyList<Nutriente> Todos =
    [
        new("energiaKcal",       "Energía",                 "kcal", 0),
        new("energiaKj",         "Energía",                 "kJ",   0),
        new("agua",              "Agua",                    "g",    1),
        new("proteinas",         "Proteínas",               "g",    1),
        new("grasaTotal",        "Grasa total",             "g",    1),
        new("carbohidratosTot",  "Carbohidratos totales",   "g",    1),
        new("carbohidratosDisp", "Carbohidratos disponibles", "g",  1),
        new("fibra",             "Fibra dietaria",          "g",    1),
        new("cenizas",           "Cenizas",                 "g",    1),
        new("calcio",            "Calcio",                  "mg",   1),
        new("fosforo",           "Fósforo",                 "mg",   1),
        new("zinc",              "Zinc",                    "mg",   2),
        new("hierro",            "Hierro",                  "mg",   2),
        new("betaCaroteno",      "β caroteno equivalentes", "µg",   1),
        new("vitaminaA",         "Vitamina A equivalentes", "µg",   1),
        new("tiamina",           "Tiamina",                 "mg",   2),
        new("riboflavina",       "Riboflavina",             "mg",   2),
        new("niacina",           "Niacina",                 "mg",   2),
        new("vitaminaC",         "Vitamina C",              "mg",   1),
        new("acidoFolico",       "Ácido fólico",            "µg",   1),
        new("sodio",             "Sodio",                   "mg",   1),
        new("potasio",           "Potasio",                 "mg",   1),
    ];

    public static readonly IReadOnlyList<string> Claves =
        Todos.Select(n => n.Clave).ToArray();

    /// <summary>
    /// Descomposición del hierro. No está en <see cref="Todos"/> porque no
    /// viene de la tabla: se calcula al sumar, repartiendo el hierro de cada
    /// alimento según su origen.
    ///
    /// La separación importa porque la biodisponibilidad del hierro hemo
    /// (15-35 %) y la del no hemo (2-20 %) no son comparables. Las
    /// preparaciones —sopas, segundos— caen en <see cref="HierroMixto"/>:
    /// contienen las dos formas y atribuirlas a una sola sería inventar.
    /// </summary>
    public const string HierroHemo = "hierroHemo";
    public const string HierroNoHemo = "hierroNoHemo";
    public const string HierroMixto = "hierroMixto";

    public const string OrigenAnimal = "Animal";
    public const string OrigenVegetal = "Vegetal";
    public const string OrigenMixto = "Mixto";

    /// <summary>Normaliza el origen a uno de los tres valores admitidos.</summary>
    public static string NormalizarOrigen(string? origen) => origen?.Trim().ToLowerInvariant() switch
    {
        "animal" => OrigenAnimal,
        "mixto" => OrigenMixto,
        _ => OrigenVegetal,
    };
}

File: frontend/src/lib/nutrientes.ts
/* ────────────────────────────────────────────────────────────────
   CATÁLOGO DE NUTRIENTES

   Réplica de `backend/src/NutriSoftware.Domain/Nutrientes.cs`. Las claves
   tienen que coincidir exactamente: son las que viajan en los JSON de
   composición, tanto al leer el catálogo de alimentos como al guardar un
   recordatorio.

   Si se agrega un nutriente, va en los dos lados.
   ──────────────────────────────────────────────────────────────── */

export interface Nutriente {
  clave: string
  etiqueta: string
  unidad: string
  decimales: number
  /** Los que se muestran en la tira de resumen sin desplegar el detalle. */
  destacado?: boolean
}

export const NUTRIENTES: Nutriente[] = [
  { clave: 'energiaKcal',       etiqueta: 'Energía',                   unidad: 'kcal', decimales: 0, destacado: true },
  { clave: 'energiaKj',         etiqueta: 'Energía',                   unidad: 'kJ',   decimales: 0 },
  { clave: 'agua',              etiqueta: 'Agua',                      unidad: 'g',    decimales: 1 },
  { clave: 'proteinas',         etiqueta: 'Proteínas',                 unidad: 'g',    decimales: 1, destacado: true },
  { clave: 'grasaTotal',        etiqueta: 'Grasa total',               unidad: 'g',    decimales: 1, destacado: true },
  { clave: 'carbohidratosTot',  etiqueta: 'Carbohidratos totales',     unidad: 'g',    decimales: 1, destacado: true },
  { clave: 'carbohidratosDisp', etiqueta: 'Carbohidratos disponibles', unidad: 'g',    decimales: 1 },
  { clave: 'fibra',             etiqueta: 'Fibra dietaria',            unidad: 'g',    decimales: 1, destacado: true },
  { clave: 'cenizas',           etiqueta: 'Cenizas',                   unidad: 'g',    decimales: 1 },
  { clave: 'calcio',            etiqueta: 'Calcio',                    unidad: 'mg',   decimales: 1 },
  { clave: 'fosforo',           etiqueta: 'Fósforo',                   unidad: 'mg',   decimales: 1 },
  { clave: 'zinc',              etiqueta: 'Zinc',                      unidad: 'mg',   decimales: 2 },
  { clave: 'hierro',            etiqueta: 'Hierro',                    unidad: 'mg',   decimales: 2, destacado: true },
  { clave: 'betaCaroteno',      etiqueta: 'β caroteno equivalentes',   unidad: 'µg',   decimales: 1 },
  { clave: 'vitaminaA',         etiqueta: 'Vitamina A equivalentes',   unidad: 'µg',   decimales: 1 },
  { clave: 'tiamina',           etiqueta: 'Tiamina',                   unidad: 'mg',   decimales: 2 },
  { clave: 'riboflavina',       etiqueta: 'Riboflavina',               unidad: 'mg',   decimales: 2 },
  { clave: 'niacina',           etiqueta: 'Niacina',                   unidad: 'mg',   decimales: 2 },
  { clave: 'vitaminaC',         etiqueta: 'Vitamina C',                unidad: 'mg',   decimales: 1 },
  { clave: 'acidoFolico',       etiqueta: 'Ácido fólico',              unidad: 'µg',   decimales: 1 },
  { clave: 'sodio',             etiqueta: 'Sodio',                     unidad: 'mg',   decimales: 1 },
  { clave: 'potasio',           etiqueta: 'Potasio',                   unidad: 'mg',   decimales: 1 },
]

export const NUTRIENTES_DESTACADOS = NUTRIENTES.filter(n => n.destacado)

/** Descomposición del hierro. No viene de la tabla: la calcula el servidor. */
export const HIERRO_DESGLOSE: Nutriente[] = [
  { clave: 'hierroHemo',   etiqueta: 'Hierro hemo (animal)',     unidad: 'mg', decimales: 2 },
  { clave: 'hierroNoHemo', etiqueta: 'Hierro no hemo (vegetal)', unidad: 'mg', decimales: 2 },
  { clave: 'hierroMixto',  etiqueta: 'Hierro sin atribuir',      unidad: 'mg', decimales: 2 },
]

export type OrigenAlimento = 'Animal' | 'Vegetal' | 'Mixto'

/**
 * Origen del hierro según la categoría del alimento.
 *
 * El Excel obligaba a marcar Animal o Vegetal a mano en cada carga, sin
 * valor por defecto y sin relación con el alimento elegido. Acá sale de la
 * categoría y se puede corregir puntualmente.
 *
 * Las preparaciones quedan en «Mixto»: un segundo con carne y arroz aporta
 * hierro hemo y no hemo a la vez, y asignarlo a uno solo sería inventar.
 */
const ORIGEN_POR_CATEGORIA: Record<string, OrigenAlimento> = {
  'Carnes y derivados':             'Animal',
  'Pescados y mariscos':            'Animal',
  'Leche y derivados':              'Animal',
  'Huevos':                         'Animal',
  'Fórmulas y papillas infantiles': 'Animal',
  'Sopas y entradas':               'Mixto',
  'Segundos y platos preparados':   'Mixto',
}

export function origenPorCategoria(categoria?: string | null): OrigenAlimento {
  if (!categoria) return 'Vegetal'
  return ORIGEN_POR_CATEGORIA[categoria] ?? 'Vegetal'
}

/**
 * Formatea un valor respetando la distinción entre «sin dato» y cero.
 * La tabla peruana tiene 5.009 celdas sin dato sobre 41.140, así que
 * mostrarlas como 0 sería afirmar algo que la tabla no dice.
 */
export function formatearNutriente(valor: number | null | undefined, n: Nutriente): string {
  if (valor === null || valor === undefined) return '—'
  return valor.toFixed(n.decimales)
}

export function buscarNutriente(clave: string): Nutriente | undefined {
  return NUTRIENTES.find(n => n.clave === clave)
    ?? HIERRO_DESGLOSE.find(n => n.clave === clave)
}
